File: ReferenceCycles/ReferenceCyclesMaster/ReferenceCyclesMaster/Person.m
//
//  Person.m
//  ReferenceCyclesMasterOc
//
//  Created by tlc on 2018/4/10.
//  Copyright © 2018年 tlc. All rights reserved.
//

#import "Person.h"

@interface Person()
{
    CallBackBlock _block;
}
@end

@implementation Person

- (instancetype)initWithName: (NSString *)name age: (NSUInteger)age {
    if (self = [super init]) {
        self.name = name;
        self.age  = age;
    }
    return self;
}

- (void)personAgeChange:(CallBackBlock)callBack {
    _block = callBack;
}

- (void)setAge:(NSUInteger)age {
    _age = age;
    
    if (_block) {
        _block();
    }
    
}

- (void)dealloc
{
    NSLog(@"person 释放");
}

@end

File: ReferenceCycles/ReferenceCyclesMaster/ReferenceCyclesMaster/Person.swift
//
//  Person.swift
//  ReferenceCyclesMaster
//
//  Created by tlc on 2018/4/9.
//  Copyright © 2018年 tlc. All rights reserved.
//

import UIKit

class Person: NSObject {
    var name: String
    var age: Int
    
    typealias PersonAgeChangedClosure = ()->String
    lazy var closure: PersonAgeChangedClosure = { [unowned self] () -> String in
//        if let strongSelf = self {
            return "姓名：\(self.name), 年龄：\(self.age)"
//        }
//        return ""
    }
    
    init(name: String, age: Int) {
        self.name = name
        self.age = age
        
        print("Person.name: \(name), age:\(age) init")
    }
    
    deinit {
        print("Person.name: \(name), age:\(age) deinit")
    }
}
